### PR TITLE
feat/view: assure heads in a view are sorted

### PR DESCRIPTION
**Pull request by drahnr**
Assure sorted heads in `View`s.

Ref #2133

### CODE REVIEW

File: node/network/protocol/src/lib.rs [hunk 1]
```diff
@@ -184,14 +185,69 @@ macro_rules! our_view {
 /// and the highest known finalized block number.
 ///
 /// Up to `N` (5?) chain heads.
-#[derive(Default, Debug, Clone, PartialEq, Eq, Encode, Decode)]
+#[derive(Default, Debug, Clone, Encode, Decode)]
 pub struct View {
 	/// A bounded amount of chain heads.
+	/// Invariant: Contains no duplicates.
```
Comment (bkchr) [inline node/network/protocol/src/lib.rs:191]: Than use a `HashSet` :shrug:

File: node/network/protocol/src/lib.rs [hunk 2]
```diff
@@ -184,14 +185,69 @@ macro_rules! our_view {
...
+			// check if the previous inner matches
+			if let Some(last_inner) = last_inner {
+				if last_inner != x {
+					return false
+				}
+			}
+			// advance inner as much as possible
+			'inner: while let Some(y) = inner.next() {
+				last_inner = Some(y);
+				if y == x {
```
Comment (bkchr) [inline node/network/protocol/src/lib.rs:233]: Why do we allow duplicate hashes?

File: node/network/protocol/src/lib.rs [hunk 3]
```diff
@@ -184,14 +185,69 @@ macro_rules! our_view {
...
+	/// `finalized_number`.
+	/// Equivalence is also.
+	pub fn is_superset_of(&self, other: &Self) -> bool {
+		// None of the others' heads must be missed in our heads.
+		// Resistant to duplicates.
+		other.heads.iter().filter(|x| !self.heads.contains(x)).count() == 0
+	}
+}
+
+impl PartialEq for View {
```
Comment (bkchr) [inline node/network/protocol/src/lib.rs:210]: I don't see any advantage of this self written `PartialEq` implementation.

Comment (drahnr) [reply node/network/protocol/src/lib.rs:210]: That's correct, given the assumptions there are no duplicates. To be removed.

File: node/network/protocol/src/lib.rs (hunk 1 above)
Comment (drahnr) [reply node/network/protocol/src/lib.rs:191]: Ahh, that's wrong. Should be `sorted`.

**Comment by ordian** [comment]
I'd be in favor of enforcing it in a type system by using `SortedVec` or `BTreeSet` or whatever. Also we can enforce heads in peer updates are sorted here https://github.com/paritytech/polkadot/blob/fc655ee5ea3322b43547f2ac0ca79976fda3cca0/node/network/bridge/src/lib.rs#L480

**Comment by ordian** [comment]
> I think we should not rely on outgoing wire messages having the correct sorting, I'd rather sort the incoming wire messages.

By enforcing it, I meant issue a malformed_view penalty if a peer sent us an unsorted view. We already do this if the update is big enough (> 5 heads).

**Comment by drahnr** [comment]
`SortedVec` is not really helpful in this context, the `View` is our container which should enforce sorting, possibly hiding the internal `Vec<Hash>` instead.

File: node/network/protocol/src/lib.rs [hunk 4]
```diff
@@ -234,6 +289,13 @@ impl View {
 	pub fn contains(&self, hash: &Hash) -> bool {
 		self.heads.contains(hash)
 	}
+
+	/// Check if two views have the same heads.
+	///
+	/// Equivalent to the
```
Comment (ordian) [inline node/network/protocol/src/lib.rs:295]: the sentence is unfinished

File: node/network/protocol/src/lib.rs [hunk 5]
```diff
@@ -206,11 +206,66 @@ pub struct View {
...
+	/// Add a head to the view.
+	///
+	/// Prevents duplicates.
+	pub fn add(&mut self, head: Hash) {
+		match { self.heads.binary_search(&head) } {
+			Ok(_pos) => {} // element already contained
+			Err(pos) => self.heads.insert(pos, head),
+		}
+	}
+
```
Comment (ordian) [inline node/network/protocol/src/lib.rs:268]: YAGNI
```suggestion
```

File: node/network/approval-distribution/src/tests.rs [hunk 6]
```diff
@@ -666,7 +666,7 @@ fn update_peer_view() {
 		overseer_send(
 			overseer,
 			ApprovalDistributionMessage::NetworkBridgeUpdateV1(
-				NetworkBridgeEvent::PeerViewChange(peer.clone(), View { heads: vec![hash_b, hash_c, hash_d], finalized_number: 2 })
+				NetworkBridgeEvent::PeerViewChange(peer.clone(), View::new(vec![hash_b, hash_c, hash_d], 2 ))
```
Comment (ordian) [inline node/network/approval-distribution/src/tests.rs:669]:
```suggestion
				NetworkBridgeEvent::PeerViewChange(peer.clone(), View::new(vec![hash_b, hash_c, hash_d], 2))
```

File: node/network/bridge/src/lib.rs [hunk 7]
```diff
@@ -1378,10 +1379,7 @@ mod tests {
 
 			let actions = network_handle.next_network_actions(2).await;
 			let wire_message = WireMessage::<protocol_v1::ValidationProtocol>::ViewUpdate(
-				View {
-					heads: vec![hash_b],
-					finalized_number: 1,
-				}
+				View::new(vec![hash_b],1)
```
Comment (ordian) [inline node/network/bridge/src/lib.rs:1382]:
```suggestion
				View::new(vec![hash_b], 1)
```

**Review by ordian — APPROVED** [review]

**Review by eskimor — APPROVED** [review]
Looks good to me.

**Comment by drahnr** [comment]
bot merge

**Pull request merged**